feat: 회원 기능, 메일 기능 구현

wndgndi opened the pull request:
### 변경사항
**AS-IS**

**TO-BE**

***Mail***
1.  `getAuthKey` : 인증키를 생성합니다.
2.  `sendEmail` : 입력받은 이메일로 인증키를 발송합니다.


***Auth***
1. ` signup` : 회원가입 기능입니다.
2. ` login` : 아이디와 패스워드가 일치하여 로그인이 성공하면 AccessToken과 RefreshToken을 발급해줍니다.
3.  `logout` : RefreshToken을 redis에서 삭제하고, AccessToken을 redis에 블랙리스트로 등록하여 로그아웃을 합니다.
4.  `reissue` : RefreshToken이 유효하다면, 새로 토큰을 발급해줍니다.
5. `validateMemberName` : 아이디가 존재하는지 검사합니다.
6. `validateEmail` : 이메일이 존재하는지 검사하고, 사용 가능하면 메일을 발송합니다.
7.  `validateAuthKey` : 입력받은 키와 인증키가 일치하는지 검사합니다.

***Member*** 
1.` deleteMember` : 회원이 탈퇴합니다.
2. `getMember` : 회원 정보를 가져옵니다.
3. `updateMember` : 회원 정보를 수정합니다. ( 아이디와 이메일을 수정할 때는 사용가능 여부를 확인합니다 )

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [ ] 테스트 코드
- [x] API 테스트

wndgndi committed:
chore: 프로젝트 설정

의존성, yml, gitignore을 설정했습니다.

wndgndi committed:
feat: 예외 처리 구현

예외 처리 시, 상태 코드와 설명을 보여줍니다.

wndgndi committed:
생성 날짜, 수정 날짜를 위한 BaseEntity

wndgndi committed:
예외 추가

wndgndi committed:
feat: Redis Config와 Util 구현

wndgndi committed:
feat: Member 관련 기능 구현

Member의 Role, domain, Controller, Service, dto, Repository를 구현하였습니다.

wndgndi committed:
feat: 메일 발송 기능 구현

wndgndi committed:
feat: PasswordEncoder Config 구현

wndgndi committed:
feat: JWT 로그인, 로그아웃, 회원가입 등 auth 기능 구현

JWT를 사용하기 위한 설정, API, 필터 기능 구현

wndgndi committed:
fix: 토큰 재발급 수정

wndgndi committed:
refactor: 사용하지 않는 import 제거, ErrorCode 수정

wndgndi committed:
feat: 회원정보 수정 기능 구현

wndgndi committed:
refactor: 사용하지 않는 import 제거

wndgndi committed:
refactor: 에러 코드 수정

eTaphee commented on src/main/java/com/personal/movie/service/MemberService.java:
```diff
@@ -0,0 +1,51 @@
+    private final MemberRepository memberRepository;
+    private final AuthService authService;
+
+    public MemberResponse deleteMember(Long memberId) {
```
회원 삭제는 관리자만 호출 가능해야 할거 같습니다!

eTaphee commented on src/main/java/com/personal/movie/service/MemberService.java:
```diff
@@ -0,0 +1,51 @@
+        return MemberResponse.fromEntity(member);
+    }
+
+    public MemberResponse getMember(Long memberId) {
```
회원 수정은 관리자 또는 본인만 호출 가능해야 할거 같아요!

wndgndi replied:
아 그렇겠네요. 감사합니다!

eTaphee commented on src/main/java/com/personal/movie/dto/request/MemberRequest.java:
```diff
@@ -0,0 +1,38 @@
+    private String phoneNumber;  // 전화번호
+
+    @Enumerated(EnumType.STRING)
+    private Role role;
```
MemberRequest를 사용하는 곳을 보니 MemberController 에서 정보를 업데이트할 떄 사용하는데 role은 어떤 경우에 사용되나요>??

wndgndi replied:
role은 업데이트 할 때 사용하지 않습니다!
MemberRequest는 회원가입, 로그인, 정보 수정 시 사용됩니다.
회원 관련 DTO를 MemberReqeust 하나만 만들어서 그런 것 같습니다.

wndgndi committed:
feat: Member API 접근 권한 설정

wndgndi committed:
feat: 회원 수정, 삭제 수정

회원 정보를 수정하거나 회원 삭제는 본인과 관리자만 사용 가능합니다.

회원 정보 수정은 다른 사람이 사용 중인 이메일만 사용이 불가능합니다.

mingming-mentor commented on src/main/java/com/personal/movie/util/RedisUtil.java:
```diff
@@ -0,0 +1,35 @@
+
+@Service
+@RequiredArgsConstructor
+public class RedisUtil {
```
util 성 클래스는 보통 static으로 선언된 메소드만 포함한 클래스를 뜻합니다. 이 경우는 RedisService가 RedisComponent(@Component)가 좀더 좋을 것 같아요

mingming-mentor commented on src/main/java/com/personal/movie/controller/AuthController.java:
```diff
@@ -0,0 +1,57 @@
+        return ResponseEntity.ok(authService.validateEmail(request));
+    }
+
+    @GetMapping("/key")
```
validate를 수행하는 것이라면 POST면 좋을 것 같아요. GET은 데이터를 가지고 오는 용도(READ) 로 생각해주시면 좋을 것 같습니다.

mingming-mentor commented on src/main/java/com/personal/movie/domain/Member.java:
```diff
@@ -0,0 +1,67 @@
+
+    private String name;   // 이름
+
+    private String email;   // 이메일
```
UK를 걸어주셔도 좋을 것 같아요

mingming-mentor commented on src/main/java/com/personal/movie/service/AuthService.java:
```diff
@@ -0,0 +1,145 @@
+
+        Authentication authentication = tokenProvider.getAuthentication(accessToken);
+
+        if (redisUtil.getData(authentication.getName()) != null) {
```
어떤 데이터를 저장하는 것인지 key앞에 prefix도 함께 붙여주시면 좋을 것 같아요. 예를 들어 위의 경우 "REFRESH_TOKEN:{authentication.getName()}" 형태가 key가 되면 어떨까요? 이러면 예를 들어 member이름을 key로 redis에 refreshToken도 저장하고 user 데이터도 저장해야할 경우가 있을때 헷갈리지 않을듯해요

mingming-mentor commented on src/main/java/com/personal/movie/service/AuthService.java:
```diff
@@ -0,0 +1,145 @@
+            redisUtil.setDataExpire(accessToken, "logout", remainingTime);
+        }
+
+        return authentication.getName() + " 님이 정상적으로 로그아웃 되었습니다.";
```
client에 내려주는 메세지는 controller layer에서 만들면 좋을 것 같아요.

mingming-mentor commented on src/main/java/com/personal/movie/service/MailService.java:
```diff
@@ -0,0 +1,76 @@
+
+            log.info("메일 전송 완료");
+        } catch (MessagingException e) {
+            e.printStackTrace();
```
log를 이용해서 warn이나 error로 찍어주시면 좋을 것 같아요

mingming-mentor reviewed (approved):
수고 많으셨습니다

wndgndi committed:
refactor: Redis Service에서 Component 로 변경

wndgndi committed:
refactor: 아이디와 이메일 UK로 설정

wndgndi committed:
refactor: Refresh Token 명시적으로 저장

1.아이디로만 Refresh Token 저장 시 헷갈릴 수 있으므로 Prefix를 붙여서 명시적으로 저장.

2. client에 내려주는 메세지 controller layer에서 만들어줌

wndgndi committed:
fix: member 권한 설정 변경

wndgndi committed:
refactor: 로그아웃 시 Access Token 명시적으로 저장

wndgndi committed:
refactor: 이메일 변경 수정

회원 정보 변경 시, 관리자가 아니며 다른 사람의 이메일과 겹칠 경우 에러 던짐

관리자가 회원의 정보를 수정해주는 경우, 해당 회원의 이메일을 수정하지 않으면 에러를 던질 수도 있으므로
